packages/react-query/src/shared/hooks/createHooksInternal.tsx

```diff
@@ -363,141 +380,124 @@ export function createRootHooks<
+
+        let shouldUpdate = false;
+        for (const key of trackedProps) {
+          if (prev[key] !== next[key]) {
+            shouldUpdate = true;
+            break;
+          }
+        }
+        if (shouldUpdate) {
+          setState(trackResult(next, addTrackedProp));
+        }
+      },
+      [addTrackedProp, trackedProps],
+    );
 
     const reset = React.useCallback((): void => {
       // unsubscribe from the previous subscription
-      currentSubscriptionRef.current?.();
+      currentSubscriptionRef.current?.unsubscribe();
 
-      updateState(getInitialState);
       if (!enabled) {
+        updateState(() => ({ ...initialStateIdle, reset }));
         return;
       }
-
-      let isStopped = false;
+      updateState(() => ({ ...initialStateConnecting, reset }));
       const subscription = client.subscription(
         path.join('.'),
         input ?? undefined,
         {
           onStarted: () => {
-            if (!isStopped) {
-              optsRef.current.onStarted?.();
-              updateState((prev) => ({
-                ...prev,
-                status: 'pending',
-                error: null,
-              }));
-            }
+            optsRef.current.onStarted?.();
+            updateState((prev) => ({
+              ...prev,
+              status: 'pending',
+              error: null,
+            }));
           },
           onData: (data) => {
-            if (!isStopped) {
-              optsRef.current.onData?.(data);
-              updateState((prev) => ({
-                ...prev,
-                status: 'pending',
-                data,
-                error: null,
-              }));
-            }
+            optsRef.current.onData?.(data);
+            updateState((prev) => ({
+              ...prev,
+              status: 'pending',
+              data,
+              error: null,
+            }));
           },
           onError: (error) => {
-            if (!isStopped) {
-              optsRef.current.onError?.(error);
-              updateState((prev) => ({
-                ...prev,
-                status: 'error',
-                error,
-              }));
-            }
+            optsRef.current.onError?.(error);
+            updateState((prev) => ({
+              ...prev,
+              status: 'error',
+              error,
+            }));
           },
           onConnectionStateChange: (result) => {
             const delta = {
               status: result.state,
               error: result.error,
             } as $Result;
 
-            updateState((prev) => {
-              return {
-                ...prev,
-                ...delta,
-              };
-            });
+            updateState((prev) => ({
+              ...prev,
+              ...delta,
+            }));
+          },
+          onComplete: () => {
+            optsRef.current.onComplete?.();
+            updateState((prev) => ({
+              ...prev,
+              status: 'idle',
```

Comment:
onComplete setting the state to idle does seem like the sort of naming inconsistency worth me questioning

Reply:
Added a comment in b069a92177168fc9f08fa150fe219015d5f6a794 

Basically, the status of the subscription is idle, but the connection might not be (in the case of WS where there's a persistent one)